python/lsst/daf/butler/core/formatter.py

```diff
@@ -385,6 +385,35 @@ def segregateParameters(self, parameters: Optional[Dict[str, Any]] = None) -> Tu
 
         return supported, unsupported
 
+    def selectResponsibleComponent(self, readComponent: str, fromComponents: Set[Optional[str]]) -> str:
+        """Given a possible set of components to choose from, return the
+        component that should be used to calculate the requested read
+        component.
+
+        Parameters
+        ----------
+        readComponent : `str`
+            The component that is being requested.
+        fromComponents : `set` of `str`
+            The available set of component options from which that read
+            component can be derived. `None` can be included but should
+            be ignored.
+
+        Returns
+        -------
+        required : `str`
+            The component that should be used.
+
+        Raises
+        ------
+        NotImplementedError
+            Raised if this formatter refuses to answer the question.
+        ValueError
+            Raised if this formatter can not determine a relevant component
+            from the supplied options.
+        """
+        raise NotImplementedError("This formatter does not support read components")
+
```

Comment:
You are absolutely right that this look up should be on the assembler and not the formatter. It is a function of the python type so the assembler does know which component is best. This could mean I can remove all the code that added the extra row although I need to ponder how the component can find out the composite storage class so it can get the right assembler. On the other hand the relationship of bbox to Exposure is something that can be worked out without caring about the formatter -- I was using the formatter because the formatter knew about the python type as well but that's clearly superfluous.  I might have to change DatasetType to add a compositeAssembler property.

This is what code review is all about! Granted it means a lot of the code on this ticket is irrelevant.